pkg/controller/installation/windows_controller.go

```diff
@@ -0,0 +1,597 @@
+			r.status.SetDegraded(operatorv1.CertificateError, "Unable to fetch prometheus certificate", err, reqLogger)
+			return reconcile.Result{}, err
+		}
+		if prometheusClientCert != nil {
+			typhaNodeTLS.TrustedBundle.AddCertificates(prometheusClientCert)
+		}
+	}
+
+	// Secure calico kube controller metrics.
+	var kubeControllerTLS certificatemanagement.KeyPairInterface
+	if instance.Spec.Variant == operatorv1.TigeraSecureEnterprise {
+		// Create or Get TLS certificates for kube controller.
+		kubeControllerTLS, err = certificateManager.GetOrCreateKeyPair(
+			r.client,
+			kubecontrollers.KubeControllerPrometheusTLSSecret,
+			common.OperatorNamespace(),
+			dns.GetServiceDNSNames(kubecontrollers.KubeControllerMetrics, common.CalicoNamespace, r.clusterDomain))
+		if err != nil {
+			r.status.SetDegraded(operatorv1.ResourceReadError, "Error finding or creating TLS certificate kube controllers metric", err, reqLogger)
+			return reconcile.Result{}, err
+		}
+
+		// Add prometheus client certificate to Trusted bundle.
+		kubecontrollerprometheusTLS, err := certificateManager.GetCertificate(r.client, monitor.PrometheusClientTLSSecretName, common.OperatorNamespace())
+		if err != nil {
+			r.status.SetDegraded(operatorv1.ResourceReadError, "Failed to get certificate for kube controllers", err, reqLogger)
+			return reconcile.Result{}, err
+		} else if kubecontrollerprometheusTLS != nil {
+			typhaNodeTLS.TrustedBundle.AddCertificates(kubeControllerTLS, kubecontrollerprometheusTLS)
+		}
+	}
+
+	components := []render.Component{}
+
+	// Fetch any existing default BGPConfiguration object.
+	bgpConfiguration := &crdv1.BGPConfiguration{}
+	err = r.client.Get(ctx, types.NamespacedName{Name: "default"}, bgpConfiguration)
+	if err != nil && !apierrors.IsNotFound(err) {
+		r.status.SetDegraded(operatorv1.ResourceReadError, "Unable to read BGPConfiguration", err, reqLogger)
+		return reconcile.Result{}, err
+	}
+
+	var hasWindowsNodes bool
+	// Don't render calico-node-windows if it's disabled in the installation
+	if instance.Spec.CalicoNetwork != nil && (instance.Spec.CalicoNetwork.WindowsDataplane == nil || *instance.Spec.CalicoNetwork.WindowsDataplane == operatorv1.WindowsDataplaneDisabled) {
+		reqLogger.Info("Calico Windows daemonset is disabled in the operator installation")
+	} else {
+		// Build a configuration for rendering calico-node-windows, but only if there are Windows nodes in the cluster
+		hasWindowsNodes, err = common.HasWindowsNodes(r.client)
+		if err != nil {
+			r.status.SetDegraded(operatorv1.ResourceReadError, "Unable to detect if there are Windows nodes present", err, reqLogger)
+		}
+		if hasWindowsNodes {
+			if len(instance.Spec.ServiceCIDRs) == 0 {
+				r.status.SetDegraded(operatorv1.ResourceValidationError, "Calico for Windows requires at least one Kubernetes service CIDR to be configured in InstallationSpec.ServiceCIDRs", fmt.Errorf("InstallationSpec.ServiceCIDRs must be configured for Calico for Windows"), reqLogger)
```

Comment:
Ah, you do have this validation present it's just not where I expected it.

I was expecting this validation to be performed based on whether or not the Installation has WindowsNodes, not whether the _cluster_ has Windows nodes. 

Those are two similar but subtly different checks. We shouldn't allow the user to specify WindowsNodes without service cidrs, but I think it's OK for a cluster to have Windows nodes without specifying WindowsNodes, right?

Reply:
Right, I think this was left over from a previous iteration in which we wouldn't have explicit windows config and thus had to detect windows nodes... Will fix, thanks for catching it!